### PR TITLE
Fix NPE in FolderActivity#onOptionsItemSelected (fixes #1190)

### LINKED ISSUE #1190: Syncthing crash in activities.FolderActivity.onOptionsItemSelected
Affected users: 10 @2018-07-25_14:43

From Gplay console, 28 minutes ago:

### Version Information
    App Version: 0.10.12
    Syncthing Version: v0.14.49
    Android Version: Android 7.0, device Huawei P9 lite (HWVNS-H)

> java.lang.NullPointerException: 
  at com.nutomic.syncthingandroid.activities.FolderActivity.onOptionsItemSelected (FolderActivity.java:440)
  at android.app.Activity.onMenuItemSelected (Activity.java:3356)
  at android.support.v4.app.FragmentActivity.onMenuItemSelected (FragmentActivity.java:380)
  at android.support.v7.app.AppCompatActivity.onMenuItemSelected (AppCompatActivity.java:195)
  at android.support.v7.view.WindowCallbackWrapper.onMenuItemSelected (WindowCallbackWrapper.java:108)
  at android.support.v7.view.WindowCallbackWrapper.onMenuItemSelected (WindowCallbackWrapper.java:108)
  at android.support.v7.app.ToolbarActionBar$2.onMenuItemClick (ToolbarActionBar.java:63)
  at android.support.v7.widget.Toolbar$1.onMenuItemClick (Toolbar.java:203)
  at android.support.v7.widget.ActionMenuView$MenuBuilderCallback.onMenuItemSelected (ActionMenuView.java:780)
  at android.support.v7.view.menu.MenuBuilder.dispatchMenuItemSelected (MenuBuilder.java:822)
  at android.support.v7.view.menu.MenuItemImpl.invoke (MenuItemImpl.java:171)
  at android.support.v7.view.menu.MenuBuilder.performItemAction (MenuBuilder.java:973)
  at android.support.v7.view.menu.MenuBuilder.performItemAction (MenuBuilder.java:963)
  at android.support.v7.widget.ActionMenuView.invokeItem (ActionMenuView.java:624)
  at android.support.v7.view.menu.ActionMenuItemView.onClick (ActionMenuItemView.java:150)
  at android.view.View.performClick (View.java:5646)
  at android.view.View$PerformClick.run (View.java:22459)
  at android.os.Handler.handleCallback (Handler.java:761)
  at android.os.Handler.dispatchMessage (Handler.java:98)
  at android.os.Looper.loop (Looper.java:156)
  at android.app.ActivityThread.main (ActivityThread.java:6523)
  at java.lang.reflect.Method.invoke (Native Method)
  at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:942)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:832)

### PR DESCRIPTION
**Pull request by ghost**
Purpose
See issue #1190 
Currently users aren't able to create new folders from the android app UI. 14.49 stuff related to the new folder types sendreceive/sendonly caused this as proper variable initialization was missing in the code in the early beginning of the app. (there was a boolean never initialized and coincidentially  the same as the xml layout's folder type default so no one noticed) With the change from boolean to string folder type, the bug hit. I'm sorry I didn't catch this earlier. If one toggles the folder type switch on the UI before saving, the bug doesn't hit. I guess the majority of users won't find this culprit by themselves.

Impact
According to GPlay Console a majority of users isn't able to create new folders and hit by this bug. So I'd request an exceptional bugfix release 0.10.13. Sorry for that :-/.
image

Testing
Verified working on my device lg-h815 running Android 7.1.2 at commit https://github.com/syncthing/syncthing-android/pull/1191/commits/d78b86f5b43eb79a4b1d0588b9f387a64b01771d .

### CODE REVIEW

**Comment by Catfriend1** [comment]
Ok wait, don't me it yet. I have a reproducer and will fix and verify this bug later. I'm after finding a time slot today...

**Comment by Catfriend1** [comment]
Investigating ...

**Comment by Catfriend1** [comment]
Ok, problem is mFolder.type == null
That is not allowed to happen. I'll fix it shortly ...

**Comment by Catfriend1** [comment]
@AudriusButkevicius Please lgtm. I hope you'll have time to make a bugfix release. I'm so sorry.

**Comment by AudriusButkevicius** [comment]
So much for testing I guess. 

I think you should start writing automated unit tests, because as we can see manual testing doesn't scale and catch everything.

**Review by AudriusButkevicius — APPROVED** [review]

**Comment by Catfriend1** [comment]
Just for reference: The problem was introduced here.
image
That's commit https://github.com/syncthing/syncthing-android/pull/1170/commits/67492bcf0bef2511a50215629c702d394264592f .
We have a problem because the UI xml said "masterFolder type default = false" and the code had nothing in it to initialize the boolean variable mFolder.type. It "java-caused" had false in it instead of null as a boolean cannot be null. When changing mFolder.type from boolean to string to prepare for syncthing version v0.14.49, it broke as the string was null (and uninitialized).

There was nothing wrong with testing as I reported my last successful tests on the PR's implementation at https://github.com/syncthing/syncthing-android/pull/1170/commits/9ee829afaf554eaf527a02a2341210f53fa5d81e . 
image

Maybe I'll write unit tests one day, but know I'm focused on coding. Perhaps anyone reading this wants to provide help on writing tests and maintaining them together with me.

**Comment by Catfriend1** [comment]
Thanks for merging it. God dammit, things got fucked up and I didn't notice.

**Comment by AudriusButkevicius** [comment]
Talking about software development in a professional fashion, I doubt you'll find a place where unit tests are optional (unless it's a complete fucking cowboy place that hires anyone and doesn't care about going out of business). Writing tests is part of a developers job.

Sure, it's open-source, we can be sloppy, but I can just not merge stuff until it has some reasonable unit test coverage to prove that it works.

I'd rather have no improvements, over improvements that might just break the app all together just because we don't feel heroic when we write tests.

**Comment by Catfriend1** [comment]
@AudriusButkevicius I see your point. It's like you stated, we are no business so if something is fucked up it gets fixed, the sooner the better! If you can't live without tests (as syncthing core has them), feel free to contribute them. This is not a valid argument putting the work on me. I think we both wouldn't be lucky if a fork has to come up to have an easier way to push app development forward and then gets out of sync so multiple app's would coexist. There are much prettier ways as making a better release and hotfix plan which you disagreed?! before. So this is my suggestion:

1) We push everything after reviews and incremental testing took place forward to the master. That's consuming a lot of time and from my side is okay as we currently handle it.
2) The master regulary gets a new beta build pushed to google play "weekly" (+/- some days okay of course as some of us might be busy)
3) The "third" beta in the cycle will be uploaded to transifex. (I assume this is roundabout the third week after the last release) People can start translating new stuff to eliminite this "german/english" artifacts in the UI we currently experience.
4) The "fourth" beta pushes a new syncthing core version in (when its ready in its regular release cycle) and gets uploaded to GPlay (most translations probably done until then) Beta, because the current handling using the release script to automatically bump syncthing core's version is not sufficient to eliminite wrapper compatibility errors before release if any come up. I guess "you" (or anyone else) didn't test the release version after syncthing 0.14.48 got 0.14.49 compiled in before pushing it to google play. It was me doing the testing on my private build server as the master wasn't even linked to 14.49-rc.4 at an pre-release point in time.
5) I the "fourth" beta doesn't turn out any high-priority issues impacting the major use of the syncthing app, it should get pushed to Google Play as a release after about two weeks.

To sum up: Approximately six weeks after the last android app version release we would finally push out a new release to google play.

If we have an urgent problem like with 0.10.12 and the folder creation, we should discuss if it's okay to do a hot-fix release only cherry-picking the most important commits from the master related to the hot-fix (as the master might have more merges in-between already).

I'm writing this in the hope "we" find an agreement that empowers the work together without creating too much overhead as we haven't much active contributing people on this repo at the moment. I would agree on putting more rules on this repo when and how changes get in when the major downsides of the app got fixed and we have more active contributors.

**Comment by AudriusButkevicius** [comment]
> This is not a valid argument putting the work on me.

So who else is supposed to prove that what you've written actually works?
Are you suggesting that you write all the features, and we get some superhero to do all the bitchwork and write the tests?

It's every man for himself, you write the features, you write tests for your own feature to prove to the community that you're not screwing things up.

It's easier to take no risk and leave things where they are, than to accept a contribution which can potentially break things.

I am not saying that that's the stance we should take, but I am saying that claiming that tests are someone else's problem is not fair on anyone.

**Comment by Catfriend1** [comment]
@AudriusButkevicius ( So who else is supposed to prove that what you've written actually works? )
My answer to this is strategy. I'm testing. Others are testing ... see my release cycle suggestion. I'm doing incremental tests after every single step during development, it's unfair to un-value this.

Leaving "things as they are" instead of constant development cause IT products to "get obsolete" and minor bugs "never" be fixed. That's also no quality assurance. What about the release script you've recently fired to update syncthing core. It's 80% of this app.
> (Quote adjusted) "It's every man for himself, you make a release, you write tests for the release to prove to the community that you're not screwing things up." You should be more thankful as I've monitored the release and impacts in the wild and reacted to the problems. I've interruped my day planning to help here.

**Comment by AudriusButkevicius** [comment]
I am not sure why you changed my quote, as unit tests are written for features/changes, not for releases.

What I am talking about is something like this:

https://github.com/syncthing/syncthing/pull/5027/files

Where the feature code is pretty much half of the pull request, the other half are tests proving that the feature works and that we're not fucking everyone over.

You could write a unit test once, and it would run automatically, without any time consumption until the end of time, not only for you, for everyone who ever wants to touch the code.

It's absurd to believe that every time you will make some change, however minor it is, you will test all the possible usage scenarios of the application manually. You simply won't as there is too many of them, or they take up too much time for what is reasonable to spend.

If we had unit tests, this potentially could have been prevented.
Not only for you, but anyone, ever. And they would have to spend pretty much no time to validate that the app still functions as it used to by simply running the tests.

I am thankful that you've addressed this issue in such a timely manner.

Again, I am not a loss if we don't have tests, nor do I strongly care, it's the users that are at a loss here.
They get angry when we break shit, and call the app shit which makes all of your effort unappreciated.

Anyways, let's agree to disagree, as I see we're not going to have any fruitful conclusions.
